src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/SyntaxNodeExtensions.cs

```diff
@@ -328,5 +329,60 @@ public static int GetTrailingWhitespaceLength(this SyntaxNode node, FormattingCo
     }
 
     public static SyntaxNode? FindInnermostNode(this SyntaxNode node, int index, bool includeWhitespace = false)
-        => node.FindToken(index, includeWhitespace)?.Parent;
+    {
+        var token = node.FindToken(index, includeWhitespace);
+
+        // If the index is EOF but the node has index-1,
+        // then try to get a token to the left of the index.
+        // patterns like
+        // <button></button>$$
+        // should get the button node instead of the razor document (which is the parent
+        // of the EOF token)
+        if (token.Kind == SyntaxKind.EndOfFile && node.Span.Contains(index - 1))
+        {
+            token = node.FindToken(index - 1, includeWhitespace);
+        }
+
+        return token.Parent;
+    }
+
+    public static SyntaxNode? FindNode(this SyntaxNode @this, Language.Syntax.TextSpan span, bool includeWhitespace = false, bool getInnermostNodeForTie = false)
+    {
+        if (!@this.FullSpan.Contains(span))
+        {
+            throw new ArgumentOutOfRangeException(nameof(span));
+        }
+
+        var node = @this.FindToken(span.Start, includeWhitespace)
+            .Parent
+            !.FirstAncestorOrSelf<SyntaxNode>(a => a.FullSpan.Contains(span));
+
+        node.AssumeNotNull();
+
+        // Tie-breaking.
+        if (!getInnermostNodeForTie)
+        {
+            var cuRoot = node.Ancestors().Last();
```

Comment:
@333fred: Should there be a cheaper way to retrieve the `RazorSyntaxTree` for a given red node?

Reply:
I was surprised there wasn't as well when I implemented `FindToken`. It's one of the things missing from our copy of Roslyn's nodes.